toolchain/parse/extract.cpp

```diff
@@ -0,0 +1,341 @@
+    if (trace) {
+      *trace << "NodeIdInCategory";
+      // TODO: Make NodeCategory printable instead.
+      if (!Category) {
+        *trace << " <none>";
+      }
+#define CARBON_NODE_CATEGORY(Name)         \
+  if (!!(Category & NodeCategory::Name)) { \
+    *trace << " " #Name;                   \
+  }
+      CARBON_NODE_CATEGORY(Decl);
+      CARBON_NODE_CATEGORY(Expr);
+      CARBON_NODE_CATEGORY(Modifier);
+      CARBON_NODE_CATEGORY(NameComponent);
+      CARBON_NODE_CATEGORY(Pattern);
+      CARBON_NODE_CATEGORY(Statement);
+#undef CARBON_NODE_CATEGORY
+    }
+
+    if (it == end || !(tree->node_kind(*it).category() & Category)) {
+      if (trace) {
+        if (it == end) {
+          *trace << " error: no more children\n";
+        } else {
+          *trace << " error: kind " << tree->node_kind(*it)
+                 << " doesn't match\n";
+        }
+      }
+      return std::nullopt;
+    }
+    if (trace) {
+      *trace << ": kind " << tree->node_kind(*it) << " consumed\n";
+    }
+    return NodeIdInCategory<Category>(*it++);
+  }
+};
+
+// Extract a `NodeIdOneOf<T, U>` as a single required child.
+template <typename T, typename U>
+struct Extractable<NodeIdOneOf<T, U>> {
+  static auto Extract(const Tree* tree, Tree::SiblingIterator& it,
+                      Tree::SiblingIterator end, ErrorBuilder* trace)
+      -> std::optional<NodeIdOneOf<T, U>> {
+    auto kind = tree->node_kind(*it);
+    if (it == end || (kind != T::Kind && kind != U::Kind)) {
+      if (trace) {
+        if (it == end) {
+          *trace << "NodeIdOneOf error: no more children, expected " << T::Kind
+                 << " or " << U::Kind << "\n";
+        } else {
+          *trace << "NodeIdOneOf error: wrong kind " << tree->node_kind(*it)
+                 << ", expected " << T::Kind << " or " << U::Kind << "\n";
+        }
+      }
+      return std::nullopt;
+    }
+    if (trace) {
+      *trace << "NodeIdOneOf " << T::Kind << " or " << U::Kind << ": "
+             << tree->node_kind(*it) << " consumed";
+    }
+    return NodeIdOneOf<T, U>(*it++);
+  }
+};
+
+// Extract a `NodeIdNot<T>` as a single required child.
+template <typename T>
+struct Extractable<NodeIdNot<T>> {
+  static auto Extract(const Tree* tree, Tree::SiblingIterator& it,
+                      Tree::SiblingIterator end, ErrorBuilder* trace)
+      -> std::optional<NodeIdNot<T>> {
+    if (it == end || tree->node_kind(*it) == T::Kind) {
+      if (trace) {
+        if (it == end) {
+          *trace << "NodeIdNot " << T::Kind << " error: no more children\n";
+        } else {
+          *trace << "NodeIdNot error: unexpected " << T::Kind << "\n";
+        }
+      }
+      return std::nullopt;
+    }
+    if (trace) {
+      *trace << "NodeIdNot " << T::Kind << ": " << tree->node_kind(*it)
+             << " consumed\n";
+    }
+    return NodeIdNot<T>(*it++);
+  }
+};
+
+// Extract a `llvm::SmallVector<T>` by extracting `T`s until we can't.
```

Comment:
Note: I have zero opinion on how best to match articles, and for `std` in particular, there's no universal pronunciation. I asked @zygoloid to give me his intended article given the different letters following and have suggested those here. Consider all of this optional. =]

```suggestion
// Extract an `llvm::SmallVector<T>` by extracting `T`s until we can't.
```